go-controller/pkg/node/node_ip_handler_linux.go

```diff
@@ -264,6 +267,21 @@ func (c *addressManager) updateNodeAddressAnnotations() error {
 	return nil
 }
 
+func (c *addressManager) updateHostAddresses(node *kapi.Node) error {
+	if config.OvnKubeNode.Mode == types.NodeModeDPU {
+		// For DPU mode, here we need to use the DPU host's IP address which is the tenant cluster's
+		// host internal IP address instead.
+		nodeAddrStr, err := util.GetNodePrimaryIP(node)
+		if err != nil {
+			return err
+		}
+		nodeAddrSet := sets.New[string](nodeAddrStr)
+		return util.SetNodeHostAddresses(c.nodeAnnotator, nodeAddrSet)
```

Comment:
It does the job. BTW, `updateNodeAddressAnnotations` is triggered by an IP address change between DPU addresses. It sounds like a side effect that we update the tenant-node annotation when the infra-node address change.

And `c.addresses` (that contains DPU addresses) is used in `addressManager.ListAddresses()`, so it's surprising why it doesn't have any other side effects.

Reply:
Hi @zeeke thank you for the review.

Do you think it would make more sense to instead put this 'OvnKubeNode.Mode == DPU' check [here ](https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/pkg/ovn/controller/services/services_controller.go#L469) and directly use the l3gateway addresses rather than updating the Host Addresses? I didn't see any other obvious place where the nodeinfo.hostaddresses field is set, but it is quite possible I missed something.

This change as is does solve the issue with NodePort service connectivity in 2-cluster design.

Reply:
@SalDaniele I think you can't check OvnKubeNode.Mode == DPU in the controller. I might be wrong, but I think OvnKubeNode.Mode is only available in the ovnkube-node.

Reply:
@zeeke In this specific case we are updating via the infra node for the tenant node. I don't believe we are able to get all the valid host-addresses from the x86 host since what ovnkube-node sees in the infra cluster is within the ARM DPU (i.e. there will be only network connections via the ports of the BF2 DPU).

Reply:
>This change as is does solve the issue with NodePort service connectivity in 2-cluster design, but we could instead set this earlier when sync() is first called so c.addresses will contain only the Node Primary IP

That makes more sense to me. Please, give it a try.

> @SalDaniele I think you can't check OvnKubeNode.Mode == DPU in the controller. I might be wrong, but I think OvnKubeNode.Mode is only available in the ovnkube-node.

True, we have to act somewhere in the ovnkube-node

> @zeeke In this specific case we are updating via the infra node for the tenant node. I don't believe we are able to get all the valid host-addresses from the x86 host since what ovnkube-node sees in the infra cluster is within the ARM DPU (i.e. there will be only network connections via the ports of the BF2 DPU).

I agree; we can't get all the valid host-addresses for the tenant node while running on DPU. I was thinking about running the `addressManager` in the  ovnkube-node-TENANT, where we can discover all the addresses:
AFAIU ovnkube-node-TENANT only runs the cni-server ([default_node_network_controller.go#L910](https://github.com/ovn-org/ovn-kubernetes/blob/37d0a302bab3f6d97d0c3660b69f06bbe5861250/go-controller/pkg/node/default_node_network_controller.go#L910)), while ovnkube-node-INFRA does all the other ovnkube-node stuff, including running the address manager. Running the address 

I understand it's a non-trivial refactor and can bring many other side effects, but it's the only way to make NodePort services listen on every x86-host interface. Maybe this scenario is not that important for DPU deployment.

Reply:
Great points, it does seem like we have addresses split between the x86-Host (tenant) and ARM-DPU (infra). Consolidating all these addresses concurrently is a change in design and a challenge for OVN-K in 2 cluster mode.
@bn222 @zshi-redhat @pliurh @DmytroLinkin @cathy-zhou @girishmg . Any options on this matter would be appreciated. The changes here are meant to support https://github.com/ovn-org/ovn-kubernetes/pull/3557 in the DPU NodePort use case.

Reply:
Let's use annotations on the DPU host side and read that. That seems to be an easy way forward.

Reply:
> > This change as is does solve the issue with NodePort service connectivity in 2-cluster design, but we could instead set this earlier when sync() is first called so c.addresses will contain only the Node Primary IP
> 
> That makes more sense to me. Please, give it a try.
> 

@zeeke AFAICT setting c.addresses to the NodePrimaryIp for DPU mode nodes solves the issue with NortPort service traffic, but results in no connectivity in clusterIP service traffic between Nodes or pod-to-pod traffic between nodes.

Reply:
@zeeke Do you know if there is a way to detect host addresses on the tenant side currently? It doesn't look like the address manager runs on tenant by default, not sure if the only option will be to enable it in order to consolidate these addresses?

Reply:
Running the address manager on the tenant side is the only way to get the tenant addresses, I think. But we can not run it on both tenant and infra, as it'd update the same node annotations.

I think we need some kind of refactoring around the addressManager to support it running on both sides.